src/mesh_smoothing.rs

```diff
@@ -91,6 +94,267 @@ pub fn laplacian_smoothing(
+            0 | 1 => (),
+            2 => {
+                // N == 2 means this is a naked edge vertex.
+                // Use (3/4, 1/8, 1/8) relocation scheme.
+
+                let vi1 = cast_usize(neighbors[0]);
+                let vi2 = cast_usize(neighbors[1]);
+
+                let v0 = geometry.vertices()[i];
+                let v1 = geometry.vertices()[vi1];
+                let v2 = geometry.vertices()[vi2];
+
+                *vertex = Point3::origin()
+                    + v0.coords * 3.0 / 4.0
+                    + v1.coords * 1.0 / 8.0
+                    + v2.coords * 1.0 / 8.0;
+            }
+            3 => {
+                // N == 3 means the vertex is not a part of a naked edge.
+                // Use (1 - N*BETA, BETA, ...) relocation scheme, where
+                // BETA is 3/16.
+
+                const N: f32 = 3.0;
+                const BETA: f32 = 3.0 / 16.0;
+
+                let vi1 = cast_usize(neighbors[0]);
+                let vi2 = cast_usize(neighbors[1]);
+                let vi3 = cast_usize(neighbors[2]);
+
+                let v1 = geometry.vertices()[vi1];
+                let v2 = geometry.vertices()[vi2];
+                let v3 = geometry.vertices()[vi3];
+
+                *vertex = Point3::origin()
+                    + vertex.coords * (1.0 - N * BETA)
+                    + v1.coords * BETA
+                    + v2.coords * BETA
+                    + v3.coords * BETA;
+            }
+            n => {
+                // N >= 3 also means the vertex is not a part of a naked edge.
+                // Use (1 - N*BETA, BETA, ...) relocation scheme, where
+                // BETA is 3 / (8*N).
+
+                let n_f32 = n as f32;
+                let beta = 3.0 / (8.0 * n_f32);
+
+                *vertex = Point3::origin() + vertex.coords * (1.0 - n_f32 * beta);
+                for vi in neighbors {
+                    let v = geometry.vertices()[cast_usize(*vi)];
+                    *vertex += v.coords * beta;
+                }
+            }
+        }
+    }
+
+    // Subdivide existing triangle faces and create new vertices
+
+    let faces_len_estimate = geometry.faces().len() * 4;
+    let mut faces: Vec<(u32, u32, u32)> = Vec::with_capacity(faces_len_estimate);
+
+    // We will be creating new mid-edge vertices per face soon. Faces
+    // will share and re-use these newly created vertices.
+
+    // The key is an unordered pair of faces that share the mid-edge
+    // vertex. The value is the index of the vertex they share.
+    let mut created_mid_vertex_indices: HashMap<UnorderedPair, u32> = HashMap::new();
+
+    for (face_index, face) in geometry.faces().iter().enumerate() {
+        let face_index_u32 = cast_u32(face_index);
+        match face {
+            Face::Triangle(triangle_face) => {
+                let (vi1, vi2, vi3) = triangle_face.vertices;
+                let face_neighbors = &face_to_face_topology[&face_index_u32];
```

Comment:
Also here you can't be sure a proper `face_to_face_topology` is passed as an argument.